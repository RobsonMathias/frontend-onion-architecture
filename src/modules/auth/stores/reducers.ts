import { combineReducers } from '@reduxjs/toolkit';
import user from './user.store';

export const rootReducer = combineReducers({
  user
});
