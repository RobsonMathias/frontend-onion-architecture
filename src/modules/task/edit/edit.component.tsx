import React, { useEffect } from 'react';

import { TaskPayload } from '@types';
import { FieldErrors, useForm } from 'react-hook-form';
import { Grid, IconButton, TextField } from '@mui/material';
import { BiCheck, BiX } from 'react-icons/bi';

export type EditComponentProps = {
  onSubmit: (payload: TaskPayload) => void;
  onCanceled: () => void;
  loading?: boolean;
  defaultValues: TaskPayload;
};

export function EditComponent(props: EditComponentProps) {
  const { onSubmit, loading, defaultValues, onCanceled } = props;

  const {
    register,
    handleSubmit,
    reset,
    formState: { errors, isSubmitted }
  } = useForm({ defaultValues });

  function hasError(error?: FieldErrors): string {
    const { message } = error || { message: '' };
    return isSubmitted ? message : '';
  }

  function submit(data: TaskPayload) {
    onSubmit(data);
    reset();
  }

  function handleCanceled() {
    onCanceled();
  }

  useEffect(() => {
    return () => {
      reset();
    };
  }, []);

  return (
    <form onSubmit={handleSubmit(submit)}>
      <Grid container>
        <TextField
          variant="filled"
          type="text"
          inputProps={{ 'data-testid': 'description' }}
          label="Insira uma atividade"
          error={!!hasError(errors.description)}
          helperText={hasError(errors.description)}
          {...register('description', {
            required: 'DescriÃ§Ã£o Ã© obrigatÃ³ria.'
          })}
          sx={{ flexGrow: 1 }}
        />
        <IconButton
          aria-label="Cancelar"
          onClick={handleCanceled}
          disabled={loading}
          data-testid="cancel"
        >
          <BiX size={24} />
        </IconButton>
        <IconButton aria-label="Adicionar" disabled={loading} data-testid="submit" type="submit">
          <BiCheck size={24} />
        </IconButton>
      </Grid>
    </form>
  );
}
