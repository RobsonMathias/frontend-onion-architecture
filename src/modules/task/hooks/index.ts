export * from './use-list';
export * from './use-task';
