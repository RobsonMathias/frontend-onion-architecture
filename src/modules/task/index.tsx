export * from './list';
export * from './new';
export * from './edit';
export * from './hooks';
