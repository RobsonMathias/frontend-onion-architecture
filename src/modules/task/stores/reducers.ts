import { combineReducers } from '@reduxjs/toolkit';
import list from './tasks.store';
import patch from './patch.store';

export const rootReducer = combineReducers({
  list,
  patch
});
