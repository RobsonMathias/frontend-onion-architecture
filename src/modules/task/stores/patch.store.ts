import { createAsyncThunk } from '@reduxjs/toolkit';
import { STORE_TASK, STORE_TASK_PATCH, STORE_STATUS_INITIALIZED } from 'constants/store';
import { ErrorStore, Task, TaskResponse, TaskStatus } from '@types';
import { createGenericSlice, createReducers, ExtraReducers, GenericState } from 'stores/generics';

import { Payload } from 'stores';

export type PatchTaskState = Task;

type State = GenericState<PatchTaskState>;

export const initialState: State = {
  data: {
    description: '',
    id: '',
    status: TaskStatus.PROGRESS
  },
  status: STORE_STATUS_INITIALIZED,
  error: null
};

export const patchTask = createAsyncThunk<PatchTaskState, Promise<TaskResponse>, Payload>(
  STORE_TASK_PATCH,
  async (promise, { rejectWithValue }) => {
    try {
      const response = await promise;
      return response;
    } catch (e: unknown) {
      const { code, message } = e as ErrorStore;
      return rejectWithValue({ code, message });
    }
  }
);

const extraReducers: ExtraReducers<PatchTaskState> = createReducers<
  PatchTaskState,
  TaskResponse,
  Payload
>(patchTask, initialState.data);

const slice = createGenericSlice({
  name: STORE_TASK,
  initialState,
  reducers: {},
  extraReducers
});

const { actions, reducer } = slice;

const { clear } = actions;

export { clear };

export default reducer;
