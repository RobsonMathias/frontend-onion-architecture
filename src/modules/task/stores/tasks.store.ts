import { createAsyncThunk } from '@reduxjs/toolkit';
import { STORE_TASK_FETCH_LIST, STORE_TASK, STORE_STATUS_INITIALIZED } from 'constants/store';
import { ErrorStore, Task, TasksResponse } from '@types';
import { createGenericSlice, GenericState, createReducers, ExtraReducers } from 'stores/generics';

import { Payload } from 'stores';

export type TasksState = Task[];

export const initialState: GenericState<TasksState> = {
  data: [],
  status: STORE_STATUS_INITIALIZED,
  error: null
};

export const fetchList = createAsyncThunk<TasksState, Promise<TasksResponse>, Payload>(
  STORE_TASK_FETCH_LIST,
  async (promise, { rejectWithValue }) => {
    try {
      const response = await promise;
      return response.data;
    } catch (e: unknown) {
      const { code, message } = e as ErrorStore;
      return rejectWithValue({ code, message });
    }
  }
);

const extraReducers: ExtraReducers<TasksState> = createReducers<TasksState, TasksResponse, Payload>(
  fetchList,
  initialState.data
);

const slice = createGenericSlice({
  name: STORE_TASK,
  initialState,
  reducers: {},
  extraReducers
});

const { actions, reducer } = slice;

const { clear } = actions;

export { clear };

export default reducer;
