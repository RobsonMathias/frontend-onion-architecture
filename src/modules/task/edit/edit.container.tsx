import React from 'react';

import { useTask } from '../hooks';
import { EditComponent } from './edit.component';
import { Task, TaskPayload } from '@types';
import { STORE_STATUS_LOADING } from 'constants/store';

export type EditContainerProps = {
  task: Task;
  onCompleted: (task?: Task) => void;
};

export function EditContainer(props: EditContainerProps) {
  const { task, onCompleted } = props;
  const { create, task: store } = useTask();

  function handleSubmit(payload: TaskPayload) {
    create(payload);
    onCompleted({ ...task, ...payload });
  }

  function handleCanceled() {
    onCompleted(undefined);
  }

  return (
    <EditComponent
      onSubmit={handleSubmit}
      onCanceled={handleCanceled}
      defaultValues={task}
      loading={store.status === STORE_STATUS_LOADING}
    />
  );
}
