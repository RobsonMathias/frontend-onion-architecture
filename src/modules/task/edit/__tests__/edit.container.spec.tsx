/**
 * @jest-environment jsdom
 */
import '@testing-library/jest-dom/extend-expect';

import { act, fireEvent, render } from '@testing-library/react';
import React from 'react';
import { EditContainer } from '../edit.container';
import { EditComponentProps } from '../edit.component';
import { Task, TaskStatus } from '@types';
import { mockTaskStore } from '@fixtures/stores/task';

const taskMock: Task = {
  description: 'description',
  status: TaskStatus.NEW,
  id: 'id'
};

function Mock({ onCanceled, onSubmit }: EditComponentProps) {
  return (
    <>
      <div data-testid="mock" onClick={() => onSubmit(taskMock)} />
      <div data-testid="mock-canceled" onClick={() => onCanceled()} />
    </>
  );
}

const mockCreate = jest.fn();
jest.mock('../../hooks', () => {
  return {
    useTask: jest.fn(() => ({
      create: jest.fn((...args) => mockCreate(...args)),
      task: mockTaskStore.task.patch
    }))
  };
});
jest.mock(__dirname + '/../edit.component', () => ({
  EditComponent: Mock
}));
const onCompleted = jest.fn();

describe('EditContainer', () => {
  const defaultProps = {
    task: taskMock,
    onCompleted
  };

  it('should render successfully', () => {
    const { container } = render(<EditContainer {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  it('should call onCompleted with a task', async () => {
    const { getByTestId } = render(<EditContainer {...defaultProps} />);

    await act(async () => {
      fireEvent.click(getByTestId('mock'));
    });

    expect(mockCreate).toHaveBeenCalledWith({
      description: 'description',
      id: 'id',
      status: 'new'
    });
    expect(onCompleted).toHaveBeenCalledWith({
      description: 'description',
      id: 'id',
      status: 'new'
    });
  });

  it('should call onCompleted with undefined value', async () => {
    const { getByTestId } = render(<EditContainer {...defaultProps} />);

    await act(async () => {
      fireEvent.click(getByTestId('mock-canceled'));
    });

    expect(mockCreate).not.toHaveBeenCalled();
    expect(onCompleted).toHaveBeenCalledWith(undefined);
  });
});
