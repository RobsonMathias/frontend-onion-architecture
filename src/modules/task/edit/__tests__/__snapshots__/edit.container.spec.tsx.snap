// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditContainer should render successfully 1`] = `
<div>
  <div
    data-testid="mock"
  />
  <div
    data-testid="mock-canceled"
  />
</div>
`;
